lib/path.nix

```diff
@@ -0,0 +1,174 @@
+  validSubpathString = value: errorPrefix:
+    if ! isString value then
+      throw "${errorPrefix}: Not a string"
+    else if value == "" then
+      throw "${errorPrefix}: The string is empty"
+    else if substring 0 1 value == "/" then
+      throw "${errorPrefix}: The string is an absolute path because it starts with `/`"
+    else true;
+
+  # Splits and normalises a subpath string into its components.
+  # Errors for ".." components and doesn't include "." components
+  splitSubpath = path: errorPrefix:
+    let
+      # Split the string into its parts using regex for efficiency. This regex
+      # matches patterns like "/", "/./", "/././", with arbitrarily many "/"s
+      # together. These are the main special cases:
+      # - Leading "./" gets split into a leading "." part
+      # - Trailing "/." or "/" get split into a trailing "." or ""
+      #   part respectively
+      #
+      # These are the only cases where "." and "" parts can occur
+      parts = split "/+(\\./+)*" path;
+
+      # `split` creates a list of 2 * k + 1 elements, containing the k +
+      # 1 parts, interleaved with k matches where k is the number of
+      # (non-overlapping) matches. This calculation here gets the number of parts
+      # back from the list length
+      # floor( (2 * k + 1) / 2 ) + 1 == floor( k + 1/2 ) + 1 == k + 1
+      partCount = length parts / 2 + 1;
+
+      # To assemble the final list of components we want to:
+      # - Skip a potential leading ".", normalising "./foo" to "foo"
+      # - Skip a potential trailing "." or "", normalising "foo/" and "foo/." to
+      #   "foo"
+      skipStart = if head parts == "." then 1 else 0;
+      skipEnd = if last parts == "." || last parts == "" then 1 else 0;
+
+      # We can now know the length of the result by removing the number of
+      # skipped parts from the total number
+      componentCount = partCount - skipEnd - skipStart;
+
+    in
+      # Special case of a single "." path component. Such a case leaves a
+      # componentCount of -1 due to the skipStart/skipEnd not verifying that
+      # they don't refer to the same character
+      if path == "." then []
+
+      # And we can use this to generate the result list directly. Doing it this
+      # way over a combination of `filter`, `init` and `tail` makes it more
+      # efficient, because we don't allocate any intermediate lists
+      else genList (index:
+        let
+          # To get to the element we need to add the number of parts we skip and
+          # multiply by two due to the interleaved layout of `parts`
+          value = elemAt parts ((skipStart + index) * 2);
+        in
+
+        # We don't support ".." components, see ./path-design.md
+        if value == ".." then
+          throw "${errorPrefix}: Path string contains contains a `..` component, which is not supported"
+        # Otherwise just return the part unchanged
```

Comment:
While I fully agree that we should discern subpaths from more general relative paths at the top-level interface, I don't think it's a good idea to encode the strictness into the lowest level.

When we discussed this earlier, you correctly said we can start out strict by throwing errors and relax later if needed.

`path.subpath.normalise` would only differ from a `path.relative.normalise` by erroring on `..` components. Is anything speaking against this:
- keep it generic. let `splitRelative` keep the `..` components (don't call it `splitSubpath`, because it does the same thing for both modulo throwing)
- keep it strict for now and let `path.relative.normalise` error if the components contain `..`:

  ```nix
  assert (any (x: x == "..") components); joinRelative components
  ```
 
  this is one full list iteration slower, but allows relaxing the `..` condition by changing one line.
-  keep it `joinRelative` as there is no special behavior for subpaths

We could even go as far as not throwing on `..` in `normalise`, and instead doing the split/check/join in `path.append` to make clear we only support unsurprising behavior at that point. In terms of migration paths there shouldn't be any difference. 

In any case that would simplify the code and make it easier to evolve.

Reply:
This sounds a bit premature, we don't even know if we need a relative normalisation yet. Personally I don't think we'll ever need it. The main argument against doing this is really the "it's slower", which I think is a strong one.

In fact, you just made me think of another optimization we could do: Check for `..` components using the regex. This saves us a whole bunch of `if` checks